### ISSUE TITLE
Performance regression: Significantly larger energy/force errors in the new code version

### ISSUE THREAD

**Issue by Xiang-LongDu** [SparkyTruck/deepmd-jax#26]
Hi,

When evaluating the energy and force predictions on the training set using a deepmd-jax model trained with the latest version, I noticed that the errors are significantly larger compared to the deepmd-tensorflow model. In contrast, a model trained with an older code version (from about a year ago) yields much smaller and more reasonable errors on the same training set.

Here are the test details:
1. Train a model using the older version (~1 year ago).
2. Train a model using the latest version.
3. Both models are trained on the same dataset.
4. Both models use identical training parameters, explicitly including `loss='l2'`.
5. Evaluate both models on the training set. The latest version shows a noticeable performance degradation.

Based on these results, I suspect that some errors might have been introduced during the recent updates.

<img width="673" height="526" alt="Image" src="https://github.com/user-attachments/assets/c771ec5e-d136-430e-8f1e-3cb332650d84" />

[example.zip](https://github.com/user-attachments/files/26961250/example.zip)

**Comment by SparkyTruck** [comment]
Hi, I’m not able to reproduce the regression with the provided dataset. Using the latest code version and the provided training script, I get an energy error of about 0.28 meV and a force error of about 0.031 eV/Å, which matches the better result shown in the plot. So far, I haven’t spotted any obvious discrepancy from the old code version.

Could you share a few more details about your setup, such as the code version, environment, and any settings you used? If the problem is still there, it would also be helpful if you could post your training log so I can compare it with mine.

**Comment by Xiang-LongDu** [comment]
Sure, I'd be happy to provide more details to help solve this issue.

### 1. Files & Logs
* **Training Log:** [output.log](https://github.com/user-attachments/files/26993261/output.log)
* **Model:** [Google Drive Link](https://drive.google.com/file/d/1A3s46ZWugoxnQWilt4On8FGNDFap3Hk5/view?usp=drive_link)

### 2. Code Check
I have verified the core scripts (`utils.py`, `train.py`, `dpmodel.py`, and `data.py`) in my local environment against the current `main` branch, and they are essentially identical. 

The **only** modification I made was adding a padding snippet to the `process_long_range_subset` function in `train.py`. This is necessary to avoid an `IndexError` when the length of `type_count` is smaller than `dplr_q_atoms`:

```python
if len(type_count) < len(dplr_q_atoms):
    padded_type_count = list(type_count) + [0] * (len(dplr_q_atoms) - len(type_count))
    type_count = tuple(padded_type_count)
```

### 3. Environment Details
Python: 3.12.12
* **Relevant Libraries:**
deepmd_jax: 0.2
e3nn-jax: 0.20.8
flax: 0.12.5
jax: 0.9.1
jax-md: 0.2.27
numpy: 2.0.0
optax: 0.2.6
scipy: 1.17.1

Hope this information helps with the debugging!

**Comment by SparkyTruck** [comment]
Thanks for sharing the log, however I still do not reproduce the error even with a fresh environment following new installation. Thanks for pointing out the type_count issue in DPLR though -- it is likely independent, but I should get it fixed. 

From the training log, it does seem that the error is bigger starting from the beginning steps, so there is surely something wrong. One potential caveat is getstat_bs=1 in the parameter used. This is not advised and risky. The default getstat_bs=64 should be good. If this is the culprit, then the behaviour is likely random (as it adds noise during initialization) and not stably reproducible (could be the reason why I cannot reproduce it), meaning that if you train it again with a different seed it may sometimes turn out well. As a side note, I generally recommend using mostly the default parameters (the only major difference you might use to improve accuracy is mp=True). When I run with the default,
train(
      model_type='energy',                  
      rcut=6.0,                                             
      train_data_path=["/tmp/issue26/data/"],
      save_path='testmodel_new.pkl',
      step=100000,                       
      loss='l2'
)
I can even get a somewhat smaller force error of 0.028 eV/Å.

**Comment by Xiang-LongDu** [comment]
Thank you for taking the time to try and reproduce this!

After doing a line-by-line comparison between the raw files of the old framework and the new framework with the help of gemini, I finally found the exact root cause of this "ghost" performance regression. It turns out to be an issue with JAX's default matrix multiplication precision (TF32) on modern NVIDIA GPUs.

In the old framework's utils.py, there was a global initialize() function that explicitly enforced FP32 precision:
```python
def initialize():
    # ...
    jax.config.update('jax_default_matmul_precision', 'float32') 
    # ...
initialize()
```

However, this initialization block was removed in the new framework.

As a result, on a newer NVIDIA architectures (e.g. A100), JAX defaults to TensorFloat-32 (TF32). While TF32 is great for standard neural networks, its 10-bit mantissa precision (the same as FP16) causes severe numerical noise when computing analytical derivatives (forces) from the energy predictions. This precision collapse during backpropagation leads to the massive RMSE errors I observed. So I think you couldn't reproduce this issue because you might be testing on a different GPU architecture.

Finally, I tested adding jax.config.update('jax_default_matmul_precision', 'float32') at the beginning of my training script with the new framework, and the training errors instantly dropped back to the normal, excellent levels I saw with the old codebase!

**Comment by SparkyTruck** [comment]
Great, I'm glad you solved it -- that is a gotcha in JAX. The previous code in initialize() has been moved to \_\_init\_\_.py -- so I think it should still be automatically set to 'float32' without having to manually set it.

**Comment by Xiang-LongDu** [comment]
Yes, you are right. It was my oversight as I forgot to check the __init__.py. Anyway, it has been a valuable experience for me to find the problem.
